feat(client): support beforeAll in DOM challenge tests

ojeytonwilliams opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or Gitpod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

The motivation for this are the very frequent test failures ([take this one, for example](https://github.com/freeCodeCamp/freeCodeCamp/actions/runs/13538538272/job/37835011943)) that we've been having. They seem to be caused by network issues and so I wanted to be able to mock `fetch` and avoid those errors.

More generally, hooks are a testing staple, so should give lesson writers more freedom and control.

Closes: https://github.com/freeCodeCamp/freeCodeCamp/issues/58954

<!-- Feel free to add any additional description of changes below this line -->

ojeytonwilliams committed:
refactor: modularize frame functions

ojeytonwilliams committed:
refactor: use createContent in test runner

That's what the user's tests will use, so our tests should too.

ojeytonwilliams committed:
feat: create plugin to parse --before-all--

ojeytonwilliams committed:
feat: add hooks to the challenge object

ojeytonwilliams committed:
feat: use hook in the client

ojeytonwilliams committed:
fix(curriculum): mock fetch when used.

ojeytonwilliams commented:
Once this goes in, https://github.com/freeCodeCamp/contribute/pull/607 can be merged.

moT01 commented:
The beforeAll is only supposed to run in the CI or e.g. pnpm test:curriculum, correct? 

Maybe I'm doing something wrong, but I'm not convinced it's working as intended. I don't see them when running the tests for step-12 in the authors page workshop...

```sh
Challenge Id being tested: 641da803d9892447d059804e

⠙ Testing
hooks
undefined

beforeAll
undefined
```

ojeytonwilliams commented:
@moT01 it's intended to run whenever the tests run - both in the browser and on the command line.

I'll look into it.

ojeytonwilliams committed:
fix: use hooks on command line

ojeytonwilliams committed:
test: more explicit tests

We're not interested in authorDataArr, we want to know if the right src
and author appear given the fetch response.

ojeytonwilliams committed:
fix: remove IIFE wrapper

We may want to declare variables in the script and the IIFE prevents
that

ojeytonwilliams committed:
chore: fix test title

ojeytonwilliams committed:
chore: update snapshot

ojeytonwilliams committed:
fix: mock more fetches

ojeytonwilliams commented:
I added more mocking for a couple of reasons. Firstly, a challenge test timed out despite not having any DOM tests (I assume it simply hung waiting for fetch to resolve). Secondly, we shouldn't be bombarding our CDN on every test.

ShaunSHamilton reviewed (approved):
:LGTM 👍 

Just FYI: For freecodecamp-os, we include `<script id="before-all"></script>`(not exactly) in the template, then inject the hook contents that way.

I find that cleaner than doing this.

ojeytonwilliams commented:
Right, that can work, but I don't think it works here. If we evaluate the code and then inject the script, it's too late to mock fetch - it will have already started fetching.

ShaunSHamilton commented:
@ojeytonwilliams Ah, yes. So, it works in fcc-os, because we serve the js, and use `src="/injected/script.js"`
